packages/apps/plugins/plugin-space/src/util.tsx

```diff
@@ -51,346 +50,497 @@ export const getSpaceDisplayName = (space: Space): string | [string, { ns: strin
-      id: space.key.toHex(),
-      label: isPersonalSpace ? ['personal space label', { ns: SPACE_PLUGIN }] : getSpaceDisplayName(space),
-      description: space.properties.description,
-      data: space,
-      icon: (props) => <Planet {...props} />,
-      ...partials,
-      properties: {
-        ...partials.properties,
-        disabled: space.state.get() === SpaceState.INACTIVE,
-        // TODO(burdon): Change to semantic classes that are customizable.
-        palette: isPersonalSpace ? 'teal' : undefined,
-        testId: isPersonalSpace ? 'spacePlugin.personalSpace' : 'spacePlugin.space',
-      },
+    manageNodes({
+      graph,
+      condition: hidden ? true : space.state.get() !== SpaceState.INACTIVE,
+      removeEdges: true,
+      nodes: [
+        {
+          id: space.key.toHex(),
+          data: space,
+          properties: {
+            ...partials,
+            label: isPersonalSpace ? ['personal space label', { ns: SPACE_PLUGIN }] : getSpaceDisplayName(space),
+            description: space.properties.description,
+            icon: (props: IconProps) => <Planet {...props} />,
+            disabled: space.state.get() === SpaceState.INACTIVE,
+            // TODO(burdon): Change to semantic classes that are customizable.
+            palette: isPersonalSpace ? 'teal' : undefined,
+            testId: isPersonalSpace ? 'spacePlugin.personalSpace' : 'spacePlugin.space',
+          },
+          edges: [[isPersonalSpace ? 'root' : SHARED, 'inbound']],
+        },
+      ],
     });
 
-    if (
-      space.state.get() === SpaceState.READY &&
-      Migrations.versionProperty &&
-      space.properties[Migrations.versionProperty] !== Migrations.targetVersion
-    ) {
-      node.addAction({
-        id: 'migrate-space',
-        label: ['migrate space label', { ns: SPACE_PLUGIN }],
-        icon: (props) => <Database {...props} />,
-        invoke: () => dispatch({ plugin: SPACE_PLUGIN, action: SpaceAction.MIGRATE, data: { space } }),
-        properties: { mainAreaDisposition: 'in-flow' },
-      });
-    }
+    manageNodes({
+      graph,
+      condition: folder instanceof Folder && (hidden ? true : space.state.get() !== SpaceState.INACTIVE),
+      removeEdges: true,
+      nodes: [
+        {
+          id: getId(SpaceAction.ADD_OBJECT),
+          data: actionGroupSymbol,
+          properties: {
+            label: ['create object group label', { ns: SPACE_PLUGIN }],
+            icon: (props: IconProps) => <Plus {...props} />,
+            disposition: 'toolbar',
+            // TODO(wittjosiah): This is currently a navtree feature. Address this with cmd+k integration.
+            // mainAreaDisposition: 'in-flow',
+            menuType: 'searchList',
+            testId: 'spacePlugin.createObject',
+          },
+          edges: [[space.key.toHex(), 'inbound']],
+        },
+      ],
+    });
 
-    if (!isPersonalSpace && space.state.get() === SpaceState.READY) {
-      node.addAction(
+    manageNodes({
```

Comment:
Very interesting API, looking forward to making use of this.

Reply:
Not thrilled with this either but it's a necessary crutch to reduce some boilerplate code right now.

We'll look to improve the DX of building out the graph by improving the api of the builders.